ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/bounded/BoundedTimedQueue.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.manager.bounded;
+
+import static java.util.Objects.requireNonNull;
+
+import tech.pegasys.pantheon.metrics.Counter;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.metrics.PantheonMetricCategory;
+import tech.pegasys.pantheon.metrics.noop.NoOpMetricsSystem;
+
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalUnit;
+import java.util.Locale;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class BoundedTimedQueue extends LinkedBlockingDeque<Runnable> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final long keepAlive;
+  private final TemporalUnit keepAliveUnit;
+  private final long evictionPollTimeout;
+  private final TimeUnit evictionPollTimeoutUnit;
+  private final MetricsSystem metricsSystem;
+  private final String metricName;
+  private final Counter totalEvictedTaskCounter;
+
+  public BoundedTimedQueue() {
+    this(
+        Default.CAPACITY,
+        Default.KEEP_ALIVE,
+        Default.KEEP_ALIVE_UNIT,
+        Default.EVICTION_POLL_TIMEOUT,
+        Default.EVICTION_POLL_TIMEOUT_UNIT,
+        Default.METRICS_SYSTEM);
+  }
+
+  public BoundedTimedQueue(
+      final int capacity,
+      final long keepAlive,
+      final TemporalUnit keepAliveUnit,
+      final long evictionPollTimeout,
+      final TimeUnit evictionPollTimeoutUnit,
+      final MetricsSystem metricsSystem) {
+    super(capacity);
+    this.keepAlive = keepAlive;
+    this.keepAliveUnit = keepAliveUnit;
+    this.evictionPollTimeout = evictionPollTimeout;
+    this.evictionPollTimeoutUnit = evictionPollTimeoutUnit;
+    this.metricsSystem = metricsSystem;
+    this.metricName = this.getClass().getSimpleName().toLowerCase(Locale.US).replace('-', '_');
+    this.totalEvictedTaskCounter =
+        this.metricsSystem.createCounter(
+            PantheonMetricCategory.EXECUTORS,
+            metricName + "_rejected_tasks_total",
+            "Total number of tasks rejected by this working queue.");
```

Comment:
We probably need to distinguish between different instances of these queues. e.g. if in the future we also used a bounded queue for other types of messages, we'd want to know how many transaction messages were dropped vs how many GET_DATA requests.

We already have a `metricName` passed into `MonitoredExecutors` which can be passed down for this.  Following the same pattern as is used there the metric probably becomes `metricName + "_dropped_tasks_total"`.

Reply:
Done.